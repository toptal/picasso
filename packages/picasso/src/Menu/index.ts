import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './Menu'

export { default } from './Menu'
export type MenuProps = OmitInternalProps<Props>
