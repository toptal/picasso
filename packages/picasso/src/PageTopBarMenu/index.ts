import type { OmitInternalProps } from '@toptal/picasso-shared'

import type { Props } from './PageTopBarMenu'

export { default } from './PageTopBarMenu'
export type TopBarMenuProps = OmitInternalProps<Props>
