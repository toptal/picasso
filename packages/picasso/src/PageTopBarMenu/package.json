{
  "name": "@toptal/picasso-page-top-bar-menu",
  "version": "1.0.0",
  "main": "index.js",
  "dependencies": {
    "@toptal/picasso-utils": "^1.0.0",
    "@toptal/picasso-user-badge": "^1.0.0",
    "@toptal/picasso-avatar": "^1.0.0",
    "@toptal/picasso-dropdown-compound": "^1.0.0",
    "@toptal/picasso-typography": "^1.0.0",
    "@toptal/picasso-page-top-bar/constants": "^1.0.0"
  }
}