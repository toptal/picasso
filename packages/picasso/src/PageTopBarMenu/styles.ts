import type { Theme } from '@material-ui/core/styles'
import { createStyles } from '@material-ui/core/styles'

export default ({ screens }: Theme) =>
  createStyles({
    avatar: {
      fontSize: '0.9rem',
    },
    contentUserBadge: {
      [screens('small', 'medium')]: {
        padding: '1em',
        zIndex: 1,
        backgroundColor: 'white',
        position: 'relative',
      },
    },
    content: {
      width: '15em',
      maxHeight: 'calc(100vh - 4.5rem)', // viewport minus header height

      [screens('small', 'medium')]: {
        width: '100vw',
        maxHeight: 'calc(100vh - 3rem)', // viewport minus header height
      },

      // override existing media-query because we have full height
      '@media screen and (max-height: 585px)': {
        maxHeight: 'calc(100vh - 4.5rem)', // viewport minus header height

        [screens('small', 'medium')]: {
          maxHeight: 'calc(100vh - 3rem)', // viewport minus header height
        },
      },
    },
    name: {
      fontWeight: 400,
      display: 'block',
    },
    truncateText: {
      maxWidth: '11.5rem',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
      textOverflow: 'ellipsis',
    },
    arrow: {
      color: 'white',

      [screens('small', 'medium')]: {
        marginLeft: '0.5rem',
      },
    },
    paper: {
      [screens('small', 'medium')]: {
        top: '2.5em !important',
      },
    },
  })
