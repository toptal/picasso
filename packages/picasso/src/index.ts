export { BaseProps } from '@toptal/picasso-shared'
export type { ColorType, SpacingType, SizeType } from '@toptal/picasso-shared'
export { AccordionCompound as Accordion } from './AccordionCompound'
export type { AccordionProps } from './Accordion'
export { default as AccountSelect } from './AccountSelect'
export type { AccountSelectProps } from './AccountSelect'
export { default as Amount } from './Amount'
export type { AmountProps } from './Amount'
export { default as Autocomplete } from './Autocomplete'
export type {
  AutocompleteProps,
  Item as AutocompleteItem,
} from './Autocomplete'
export { AvatarCompound as Avatar } from './AvatarCompound'
export type { AvatarProps } from './Avatar'
export { ButtonCompound as Button } from './ButtonCompound'
export type { ButtonProps, VariantType as ButtonVariantType } from './Button'
export type { ButtonCheckboxProps } from './ButtonCheckbox'
export type { ButtonRadioProps } from './ButtonRadio'
export { CheckboxCompound as Checkbox } from './CheckboxCompound'
export type { CheckboxProps } from './Checkbox'
export type { CheckboxGroupProps } from './CheckboxGroup'
export { default as Container } from './Container'
export type { ContainerProps } from './Container'
export { DropdownCompound as Dropdown } from './DropdownCompound'
export type { DropdownProps } from './Dropdown'
export { default as DropdownArrow } from './DropdownArrow'
export type { DropdownArrowProps } from './DropdownArrow'
export { default as EnvironmentBanner } from './EnvironmentBanner'
export type { EnvironmentBannerProps } from './EnvironmentBanner'
export { default as FileInput } from './FileInput'
export type { FileInputProps } from './FileInput'
export { default as FileList } from './FileList'
export type { FileListProps } from './FileList'
export { FormCompound as Form } from './FormCompound'
export type { FormProps } from './Form'
export { default as FormHint } from './FormHint'
export type { FormHintProps } from './FormHint'
export { default as FormError } from './FormError'
export type { FormErrorProps } from './FormError'
export { default as FormLabel } from './FormLabel'
export type { FormLabelProps, RequiredDecoration } from './FormLabel'
export { GridCompound as Grid } from './GridCompound'
export type { GridProps, GridSize } from './Grid'
export { HelpboxCompound as Helpbox } from './HelpboxCompound'
export type { HelpboxProps } from './Helpbox'
export { default as Image } from './Image'
export type { ImageProps } from './Image'
export { default as Indicator } from './Indicator'
export type { IndicatorProps } from './Indicator'
export { default as Input } from './Input'
export type { InputProps } from './Input'
export { default as InputAdornment } from './InputAdornment'
export type { InputAdornmentProps } from './InputAdornment'
export { TagCompound as Tag } from './TagCompound'
export type { TagProps } from './Tag'
export type { TagRectangularProps } from './TagRectangular'
export { default as Link } from './Link'
export type { LinkProps } from './Link'
export { default as Loader } from './Loader'
export type { LoaderProps } from './Loader'
export { default as Logo } from './Logo'
export type { LogoProps } from './Logo'
export { MenuCompound as Menu } from './MenuCompound'
export type { MenuProps } from './Menu'
export { default as MenuItem } from './MenuItem'
export type { MenuItemProps } from './MenuItem'
export { ModalCompound as Modal } from './ModalCompound'
export type { ModalProps } from './Modal'
export { default as MonthSelect } from './MonthSelect'
export type { MonthSelectProps } from './MonthSelect'
export { NotificationCompound as Notification } from './NotificationCompound'
export type { NotificationProps } from './Notification'
export { default as NotificationActions } from './NotificationActions'
export type { NotificationActionsProps } from './NotificationActions'
export { default as NumberInput } from './NumberInput'
export type { NumberInputProps } from './NumberInput'
export { PageCompound as Page } from './PageCompound'
export type { PageProps } from './Page'
export { default as Pagination } from './Pagination'
export type { PaginationProps } from './Pagination'
export { default as Paper } from './Paper'
export type { PaperProps } from './Paper'
export { default as PromptModal } from './PromptModal'
export type { PromptModalProps } from './PromptModal'
export { RadioCompound as Radio } from './RadioCompound'
export type { RadioProps } from './Radio'
export type { RadioGroupProps } from './RadioGroup'
export { default as RichTextEditor } from './RichTextEditor'
export type {
  RichTextEditorProps,
  RichTextEditorChangeHandler,
} from './RichTextEditor'
export { default as Select } from './Select'
export type { SelectProps } from './Select'
export type {
  Option as SelectOption,
  ValueType as SelectValueType,
} from './SelectBase'
export type { PageSidebarProps } from './PageSidebar'
export { default as Slider } from './Slider'
export type { SliderProps } from './Slider'
export { default as ShowMore } from './ShowMore'
export type { ShowMoreProps } from './ShowMore'
export { StepperCompound as Stepper } from './StepperCompound'
export type { StepperProps } from './Stepper'
export type { StepperVerticalProps } from './StepperVertical'
export { TableCompound as Table } from './TableCompound'
export type { TableProps } from './Table'
export { default as TableFooter } from './TableFooter'
export type { TableFooterProps } from './TableFooter'
export { TabsCompound as Tabs } from './TabsCompound'
export type { TabsProps } from './Tabs'
export { TagSelectorCompound as TagSelector } from './TagSelectorCompound'
export type { TagSelectorProps } from './TagSelector'
export { default as Tooltip } from './Tooltip'
export type { TooltipProps } from './Tooltip'
export { default as Typography } from './Typography'
export type { TypographyProps } from './Typography'
export { default as UserBadge } from './UserBadge'
export type { UserBadgeProps } from './UserBadge'
export { default as YearSelect } from './YearSelect'
export type { YearSelectProps } from './YearSelect'
export { default as ApplicationUpdateNotification } from './ApplicationUpdateNotification'
export type { ApplicationUpdateNotificationProps } from './ApplicationUpdateNotification'
export { AlertCompound as Alert } from './AlertCompound'
export type { AlertProps } from './Alert'
export type { AlertInlineProps } from './AlertInline'
export { default as Switch } from './Switch'
export type { SwitchProps } from './Switch'
export { default as Quote } from './Quote'
export type { QuoteProps } from './Quote'
export { default as Rating } from './Rating'
export type { RatingStarsProps, RatingThumbsProps } from './Rating'
export {
  default as TreeView,
  TreeViewContainer,
  useTreeView,
  TreeNodeAvatar,
} from './TreeView'
export type { TreeNodeInterface } from './TreeView'
export { default as StaticTreeView } from './StaticTreeView'
export { default as TimePicker } from './TimePicker'
export type { TimePickerProps } from './TimePicker'
export { default as SkeletonLoader } from './SkeletonLoader'
export { default as PageHead } from './PageHead'
export type { PageHeadProps } from './PageHead'
export { default as ProgressBar } from './ProgressBar'
export type { ProgressBarProps } from './ProgressBar'
export { default as List } from './List'
export type { ListProps } from './List'
export { default as Badge } from './Badge'
export type { BadgeProps } from './Badge'
export { BreadcrumbsCompound as Breadcrumbs } from './BreadcrumbsCompound'
export type { BreadcrumbsProps } from './Breadcrumbs'
export { default as Calendar } from './Calendar'
export type { DateOrDateRangeType } from './Calendar'
export {
  default as DatePicker,
  datePickerParseDateString,
  DEFAULT_DATE_PICKER_DISPLAY_DATE_FORMAT,
  DEFAULT_DATE_PICKER_EDIT_DATE_FORMAT,
} from './DatePicker'
export type {
  DatePickerValue,
  DatePickerInputCustomValueParser,
  DatePickerInputCustomValueParserParameters,
  DatePickerProps,
} from './DatePicker'
export { default as Drawer } from './Drawer'
export type { DrawerProps } from './Drawer'
export { default as Dropzone, ErrorCode as DropzoneErrorCode } from './Dropzone'
export type {
  DropzoneProps,
  FileUpload as DropzoneFileUpload,
  FileError as DropzoneFileError,
  FileRejection as DropzoneFileRejection,
  DropEvent as DropzoneDropEvent,
} from './Dropzone'
export { OverviewBlockCompound as OverviewBlock } from './OverviewBlockCompound'
export type { OverviewBlockProps } from './OverviewBlock'
export { default as TypographyOverflow } from './TypographyOverflow'
export type { TypographyOverflowProps } from './TypographyOverflow'
export { default as EmptyState } from './EmptyState'
export { default as Section } from './Section'
export type { SectionProps } from './Section'
export { NoteCompound as Note } from './NoteCompound'
export type { NoteProps } from './Note'
export { default as Timeline } from './Timeline'
export type { TimelineProps } from './Timeline'
export { default as RichText } from './RichText'
export type { RichTextProps, ASTType } from './RichText'
export { default as PasswordInput } from './PasswordInput'
export type { PasswordInputProps } from './PasswordInput'
export { default as FieldRequirements } from './FieldRequirements'
export type {
  FieldRequirementsProps,
  FieldRequirement,
} from './FieldRequirements'
export type { Status as OutlinedInputStatus } from './OutlinedInput'
export { default as AvatarUpload } from './AvatarUpload'
export type {
  AvatarUploadProps,
  FileUpload as AvatarUploadFileUpload,
  FileError as AvatarUploadFileError,
  FileRejection as AvatarUploadFileRejection,
  DropEvent as AvatarUploadDropEvent,
} from './AvatarUpload'

// hygen code generator inserts export statements above this comment.
export * from './Icon'
