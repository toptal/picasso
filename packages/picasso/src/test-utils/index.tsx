import type { ReactElement } from 'react'
import React from 'react'
import type { RenderOptions, RenderResult } from '@testing-library/react'
import { render } from '@testing-library/react'

import TestingPicasso from './TestingPicasso'

export type PicassoConfig = {
  titleCase: boolean
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  picassoConfig?: PicassoConfig
): RenderResult => {
  const { rerender, ...rest } = render(
    <TestingPicasso titleCase={picassoConfig?.titleCase}>{ui}</TestingPicasso>,
    options
  )

  const customRerender: typeof rerender = newUi =>
    rerender(
      <TestingPicasso titleCase={picassoConfig?.titleCase}>
        {newUi}
      </TestingPicasso>
    )

  return { rerender: customRerender, ...rest }
}

export * from '@testing-library/react'
export { customRender as render, TestingPicasso }
export { getCheckpoints } from './get-checkpoints'
