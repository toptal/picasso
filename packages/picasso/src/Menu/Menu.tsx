import type { HTMLAttributes } from 'react'
import React, { forwardRef } from 'react'
import cx from 'classnames'
import { MenuList as MUIMenuList } from '@material-ui/core'
import type { Theme } from '@material-ui/core/styles'
import { makeStyles } from '@material-ui/core/styles'
import type { BaseProps } from '@toptal/picasso-shared'

import { BackMinor16 } from '../Icon'
import MenuItem from '../MenuItem'
import Typography from '../Typography'
import { useMenu } from './hooks'
import MenuContext from './MenuContext'
import styles from './styles'
import type { MenuVariant } from './types'

export interface Props extends BaseProps, HTMLAttributes<HTMLUListElement> {
  /** Switches between slide and drilldown variants */
  variant?: MenuVariant
  /** Whether or not to handle nested navigation */
  allowNestedNavigation?: boolean
  testIds?: {
    menuItem?: string
  }
}

const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoMenu',
})

export const Menu = forwardRef<HTMLUListElement, Props>(function Menu(
  props,
  ref
) {
  const {
    children,
    className,
    style,
    variant,
    allowNestedNavigation,
    testIds,
    ...rest
  } = props
  const classes = useStyles()
  const { context, innerMenu, hasBackButton } = useMenu({ variant })
  const { onBackClick, onMenuMouseLeave } = context

  return (
    <MenuContext.Provider value={context}>
      {innerMenu ?? (
        <MUIMenuList
          {...rest}
          ref={ref}
          className={cx(classes.root, className)}
          style={style}
          onMouseLeave={onMenuMouseLeave}
        >
          {hasBackButton && allowNestedNavigation && (
            <MenuItem
              key='back'
              data-testid={testIds?.menuItem}
              onClick={onBackClick}
            >
              <Typography size='xsmall' color='dark-grey' variant='body'>
                <BackMinor16 className={classes.backButtonIcon} />
                Back
              </Typography>
            </MenuItem>
          )}
          {children}
        </MUIMenuList>
      )}
    </MenuContext.Provider>
  )
})

Menu.defaultProps = {
  variant: 'slide',
  allowNestedNavigation: true,
}

Menu.displayName = 'Menu'

export default Menu
