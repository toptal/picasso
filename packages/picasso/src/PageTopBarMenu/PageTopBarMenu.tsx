/* eslint-disable complexity */
import React, { forwardRef, ReactNode, HTMLAttributes } from 'react'
import cx from 'classnames'
import { makeStyles, Theme } from '@material-ui/core/styles'
import { BaseProps } from '@toptal/picasso-shared'

import { useBreakpoint } from '../utils'
import UserBadge from '../UserBadge'
import Avatar from '../Avatar'
import { DropdownCompound as Dropdown } from '../DropdownCompound'
import Typography from '../Typography'
import styles from './styles'

const useStyles = makeStyles<Theme>(styles, {
  name: 'PicassoTopBarMenu',
})

export interface Props extends BaseProps, HTMLAttributes<HTMLDivElement> {
  /** User full name to display */
  name: string
  /** Additional content */
  meta?: ReactNode
  /** Photo url or custom Avatar component */
  avatar?: ReactNode
  /** Menu content */
  children: ReactNode
}

export const PageTopBarMenu = forwardRef<HTMLDivElement, Props>(
  function PageTopBarMenu(props, ref) {
    const { name, meta, avatar, className, style, children, ...rest } = props
    const classes = useStyles()

    const isCompactLayout = useBreakpoint(['small', 'medium'])

    const metaContent =
      typeof meta === 'string' ? (
        <Typography
          className={classes.truncateText}
          invert={!isCompactLayout}
          size='xsmall'
        >
          {meta}
        </Typography>
      ) : (
        meta
      )

    const content = isCompactLayout ? (
      <>
        <UserBadge
          center
          size='xxsmall'
          classes={{
            root: classes.contentUserBadge,
            avatar: classes.avatar,
            name: cx(classes.name, classes.truncateText),
          }}
          name={name}
          avatar={avatar}
        >
          {meta && metaContent}
        </UserBadge>
        {children}
      </>
    ) : (
      children
    )

    const trigger = isCompactLayout ? (
      <Avatar
        size='xxsmall'
        className={classes.xsmall}
        name={name}
        src={avatar as string}
      />
    ) : (
      <UserBadge
        invert
        center
        size='xxsmall'
        classes={{
          name: cx(classes.name, classes.truncateText),
        }}
        name={name}
        avatar={avatar}
      >
        {meta && metaContent}
      </UserBadge>
    )

    return (
      <Dropdown
        {...rest}
        ref={ref}
        className={cx(classes.root, className)}
        classes={{ content: classes.content }}
        style={style}
        content={content}
        offset={{ top: isCompactLayout ? 0.8 : 'xsmall' }}
        popperOptions={{
          modifiers: {
            flip: { enabled: false },
            preventOverflow: {
              padding: 0,
            },
          },
        }}
      >
        {trigger}
        <Dropdown.Arrow className={classes.arrow} />
      </Dropdown>
    )
  }
)

PageTopBarMenu.defaultProps = {}

PageTopBarMenu.displayName = 'PageTopBarMenu'

export default PageTopBarMenu
