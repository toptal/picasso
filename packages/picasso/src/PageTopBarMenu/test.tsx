import React from 'react'
/* eslint-disable-next-line */
import { render } from '@toptal/picasso/test-utils'
import { OmitInternalProps } from '@toptal/picasso-shared'

import Menu from '../Menu'
import PageTopBarMenu, { Props } from './PageTopBarMenu'

const TestPageTopBarMenu = ({
  children,
  name,
  avatar,
}: OmitInternalProps<Props>) => {
  return (
    <PageTopBarMenu name={name} avatar={avatar}>
      {children}
    </PageTopBarMenu>
  )
}

describe('PageTopBarMenu', () => {
  it('renders', () => {
    const { container } = render(
      <TestPageTopBarMenu
        name='Jacqueline Roque'
        avatar='./jacqueline-with-flowers-1954-square.jpg'
      >
        <Menu>
          <Menu.Item>My Account</Menu.Item>
          <Menu.Item>Log Out</Menu.Item>
        </Menu>
      </TestPageTopBarMenu>
    )

    expect(container).toMatchSnapshot()
  })
})
