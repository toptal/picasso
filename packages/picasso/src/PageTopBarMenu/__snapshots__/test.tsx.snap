// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTopBarMenu renders 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="PicassoDropdown-root"
    >
      <div
        class="PicassoDropdown-anchor"
      >
        <div
          class="PicassoContainer-centerAlignItems PicassoContainer-flex PicassoUserBadge-root"
        >
          <div
            class="PicassoAvatarWrapper-root PicassoAvatarWrapper-sizeXxsmallSquare PicassoAvatarWrapper-cornerXxsmall"
          >
            <img
              alt="Jacqueline Roque"
              class="PicassoImage-root PicassoImageAvatar-image PicassoUserBadge-avatar"
              src="./jacqueline-with-flowers-1954-square.jpg"
            />
          </div>
          <div
            class="PicassoContainer-flex PicassoContainer-column PicassoUserBadge-infoContainer"
          >
            <div
              class="PicassoContainer-centerAlignItems PicassoContainer-flex"
            >
              <h4
                class="MuiTypography-root PicassoTypography-invert PicassoUserBadge-name PicassoTopBarMenu-name PicassoTopBarMenu-truncateText MuiTypography-h4 MuiTypography-displayInline"
              >
                Jacqueline Roque
              </h4>
            </div>
          </div>
        </div>
        <span
          class="PicassoDropdownArrow-root PicassoTopBarMenu-arrow PicassoDropdownArrow-medium"
        />
      </div>
    </div>
  </div>
</div>
`;
